src/libraries/System.Private.CoreLib/generators/EventSourceGenerator.cs

```diff
@@ -32,69 +32,53 @@ public partial class EventSourceGenerator : ISourceGenerator
         //         }
         //     }
 
-        public void Initialize(GeneratorInitializationContext context)
-            => context.RegisterForSyntaxNotifications(() => new SyntaxReceiver());
-
-        public void Execute(GeneratorExecutionContext context)
+        public void Initialize(IncrementalGeneratorInitializationContext context)
         {
-            SyntaxReceiver? receiver = context.SyntaxReceiver as SyntaxReceiver;
-            if ((receiver?.CandidateClasses?.Count ?? 0) == 0)
-            {
-                // nothing to do yet
-                return;
-            }
+            IncrementalValuesProvider<EventSourceClass> eventSourceClasses =
+                context.SyntaxProvider
+                .CreateSyntaxProvider(static (x, _) => IsSyntaxTargetForGeneration(x), (x, _) => GetSemanticTargetForGeneration(x))
+                .Where(x => x is not null)
+                .Collect()
+                .Combine(context.CompilationProvider)
```

Comment:
@chsienki can correct me if I'm wrong, but I believe you'll need to do something like:
https://github.com/dotnet/runtime/blob/6c50d9fad7217b0eeed2b27ca6ab9ab34bc0dd1f/src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.cs#L40-L45
to avoid causing regeneration to occur on _any_ change to the compilation.

Reply:
It was not as simple since we had to compare an immutable array of class declerations, but done.